Dockerfile

```diff
@@ -18,117 +20,129 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
     curl \
     kdiff3 \
     git \
-    vim # for debug
+    vim \
+  && rm -rf /var/lib/apt/lists/*
 
 # Install runtime packages.
-RUN apt-get update && \
-    apt-get install -y \
-    ruby \
-    python \
-    python-dev \
-    python-pip \
+RUN apt-get update \
+  && apt-get install -y \
     unzip \
     perl \
     software-properties-common \
-    php-pear
+    php-pear \
+    python3-pip \
+  && rm -rf /var/lib/apt/lists/*
+
+# Install protoc 3.3.0.
+# Currently, the stable version is still 2.6.x, which can only handle proto2
+# syntax, so we have to download our own.
+RUN mkdir -p /usr/src/protoc/ \
+  && curl --location https://github.com/google/protobuf/releases/download/v3.3.0/protoc-3.3.0-linux-x86_64.zip > /usr/src/protoc/protoc-3.3.0.zip \
+  && cd /usr/src/protoc/ \
+  && unzip protoc-3.3.0.zip \
+  && ln -s /usr/src/protoc/bin/protoc /usr/local/bin/protoc
+
+# Install GRPC and Protobuf.
+RUN pip3 install --upgrade pip \
+  && pip3 install \
+    grpcio==1.3.0 \
+    protobuf==3.3.0
 
 # Install Oracle JDK 8
-RUN add-apt-repository ppa:openjdk-r/ppa
-RUN apt-get update && apt-get install -y openjdk-8-jdk
+RUN add-apt-repository ppa:openjdk-r/ppa \
+  && apt-get update \
+  && apt-get install -y openjdk-8-jdk \
+  && rm -rf /var/lib/apt/lists/*
 
 # Setup JAVA_HOME, this is useful for docker commandline
 ENV JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64/
 RUN export JAVA_HOME
 
-# Install NodeJS
-ADD ./scripts /scripts
-WORKDIR /scripts
-# This is need to install nodejs 4.x otherwise nodejs 0.x will be installed.
-RUN bash setup_node4.sh
-RUN apt-get install -y nodejs
-
-# Install linuxbrew.
-RUN useradd -m -s /bin/bash linuxbrew
-RUN echo 'linuxbrew ALL=(ALL) NOPASSWD:ALL' >>/etc/sudoers
-USER linuxbrew
-WORKDIR /home/linuxbrew
-ENV PATH /home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:$PATH
-ENV SHELL /bin/bash
-RUN yes |ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install)"
-RUN brew doctor || true
-
-# Manually install protobuf with --no-binary flag before installing protoc
-# Python and pip are installed by 'brew install --with-plugins grpc'. We need
-# to install them preemptively, and then install the protobuf package with
-# the --no-binary flag to work around https://github.com/google/protobuf/issues/2895
-RUN brew install python
-RUN pip install --upgrade pip
-RUN pip install protobuf==3.2.0 --no-binary protobuf
-
-# Install protoc and grpc.
-RUN brew tap grpc/grpc
-RUN brew install --with-plugins grpc
-
-# Install Go1.6 from linuxbrew.
-RUN brew install go
-
-USER root
+# Install NodeJS.
+# This installs Node 4 on Ubuntu 16.04.
+RUN apt-get update \
+  && apt-get install -y \
+    nodejs \
+    npm \
+  && rm -rf /var/lib/apt/lists/* \
+  && ln -s /usr/bin/nodejs /usr/local/bin/node
```

Comment:
This `ln -s` is because the Node.js package on apt is quirky and uses `nodejs` instead of `node` as the executable.